storm-core/src/jvm/backtype/storm/scheduler/TopologyDetails.java

```diff
@@ -79,11 +96,304 @@ public StormTopology getTopology() {
+                    if (bolt.getKey().equals(anExecutorToComponent.getValue())) {
+                        _resourceList.put(anExecutorToComponent.getKey(), topology_resources);
+                    }
+                }
+            }
+        } else {
+            LOG.warn("Topology " + topologyId + " does not seem to have any bolts!");
+        }
+        // Extract spout memory info
+        if (this.topology.get_spouts() != null) {
+            for (Map.Entry<String, SpoutSpec> spout : this.topology.get_spouts().entrySet()) {
+                Map<String, Double> topology_resources = backtype.storm.scheduler.resource.Utils.parseResources(spout
+                        .getValue().get_common().get_json_conf());
+                backtype.storm.scheduler.resource.Utils.checkIntialization(topology_resources, spout.getValue().toString(), this.topologyConf);
+                for (Map.Entry<ExecutorDetails, String> anExecutorToComponent : executorToComponent.entrySet()) {
+                    if (spout.getKey().equals(anExecutorToComponent.getValue())) {
+                        _resourceList.put(anExecutorToComponent.getKey(), topology_resources);
+                    }
+                }
+            }
+        } else {
+            LOG.warn("Topology " + topologyId + " does not seem to have any spouts!");
+        }
+        //schedule tasks that are not part of components returned from topology.get_spout or topology.getbolt (AKA sys tasks most specifically __acker tasks)
+        for(ExecutorDetails exec : this.getExecutors()) {
+            if (_resourceList.containsKey(exec) == false) {
+                LOG.info(
+                        "Scheduling {} {} with memory requirement as 'on heap' - {} and 'off heap' - {} and CPU requirement as {}",
+                        this.getExecutorToComponent().get(exec),
+                        exec,
+                        this.topologyConf.get(Config.TOPOLOGY_COMPONENT_RESOURCES_ONHEAP_MEMORY_MB),
+                        this.topologyConf.get(Config.TOPOLOGY_COMPONENT_RESOURCES_OFFHEAP_MEMORY_MB),
+                        this.topologyConf.get(Config.TOPOLOGY_COMPONENT_CPU_PCORE_PERCENT));
+                this.addDefaultResforExec(exec);
+            } 
+        }
+    }
+
+    private List<ExecutorDetails> componentToExecs(String comp) {
+        List<ExecutorDetails> execs = new ArrayList<>();
+        for (Map.Entry<ExecutorDetails, String> entry : executorToComponent.entrySet()) {
+            if (entry.getValue().equals(comp)) {
+                execs.add(entry.getKey());
+            }
+        }
+        return execs;
+    }
+
+    /**
+     * Returns a representation of the non-system components of the topology graph
+     * Each Component object in the returning map is populated with the list of its
+     * parents, children and execs assigned to that component.
+     * @return a map of components
+     */
+    public Map<String, RAS_Component> getRAS_Components() {
```

Comment:
I'm not sure that RAS_Component is the right name for this class or this method.  The class has nothing to do with resources at all.  Perhaps we can call it a Component and the method getComponentGraph instead?